File: quantmind-api/agent-fastapi/app/api/dtos.py
"""Internal API request/response DTOs.

File: app/api/dtos.py
Author: Sunil+Ai Assistant
Date: 2026-09-07
Description:
    Request and response models for the FastAPI internal surface the BFF
    calls. Strict validation (extra='forbid') and the ticker allowlist
    regex are applied here as defense-in-depth behind the NestJS boundary.

Source: specs/arch/SDD.md §2.1, specs/arch/DDD.md §1.1
"""
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

_TICKER_RE = r"^[A-Z.\-]{1,10}$"


class NewsCheckRequest(BaseModel):
    """FR1 request body."""

    model_config = ConfigDict(extra="forbid")

    symbol: str = Field(pattern=_TICKER_RE)
    simulate_empty: bool = False


class ScanRequest(BaseModel):
    """FR2 request body (universe optional; server default when omitted)."""

    model_config = ConfigDict(extra="forbid")

    universe: list[str] | None = None


class StrategyRequest(BaseModel):
    """FR3/FR4 request body."""

    model_config = ConfigDict(extra="forbid")

    symbol: str = Field(pattern=_TICKER_RE)
    user_framing: str | None = Field(default=None, max_length=500)
    news_label: str | None = None


class ErrorResponse(BaseModel):
    """Standard sanitized error envelope (no stack, no internal ids)."""

    model_config = ConfigDict(extra="forbid")

    code: str
    message: str
    correlation_id: str
    details: dict = Field(default_factory=dict)

File: valtide-api/bff-nestjs/src/features/dtos.ts
/**
 * File: src/features/dtos.ts
 * Author: Sunil+Ai Assistant
 * Date: 2026-09-07
 * Description:
 *   Client-facing request DTOs validated by the global ValidationPipe
 *   (class-validator) at the BFF boundary. The ticker allowlist regex is
 *   enforced here BEFORE any downstream call (STOCK-101-T2). Swagger
 *   decorators document the contract for /api/docs.
 * Source: specs/arch/SDD.md §2.1, specs/arch/DDD.md §1.1
 */
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import {
  IsArray,
  IsBoolean,
  IsOptional,
  IsString,
  Matches,
  MaxLength,
} from 'class-validator';

/** Ticker allowlist: 1-10 uppercase letters, dot, or hyphen. */
const TICKER_REGEX = /^[A-Z.\-]{1,10}$/;

/** FR1 news-check request. */
export class NewsCheckDto {
  @ApiProperty({ example: 'NVDA', description: 'Ticker symbol (allowlisted).' })
  @Matches(TICKER_REGEX, { message: 'symbol must match ^[A-Z.\\-]{1,10}$' })
  symbol!: string;

  @ApiPropertyOptional({ description: 'Force empty headline set (edge-case demo).' })
  @IsOptional()
  @IsBoolean()
  simulateEmpty?: boolean;
}

/** FR2 scan request (universe optional; server default when omitted). */
export class ScanDto {
  @ApiPropertyOptional({ type: [String], description: 'Explicit universe override.' })
  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  universe?: string[];
}

/** FR3/FR4 strategy request. */
export class StrategyDto {
  @ApiProperty({ example: 'NVDA', description: 'Ticker symbol (allowlisted).' })
  @Matches(TICKER_REGEX, { message: 'symbol must match ^[A-Z.\\-]{1,10}$' })
  symbol!: string;

  @ApiPropertyOptional({ description: 'Free-text framing (neutralized server-side).' })
  @IsOptional()
  @IsString()
  @MaxLength(500)
  userFraming?: string;

  @ApiPropertyOptional({ description: 'Originating FR1 news label, if any.' })
  @IsOptional()
  @IsString()
  newsLabel?: string;
}
